lib/swim.ml

```diff
@@ -0,0 +1,127 @@
+open Base.Hashtbl
+
+type message =
+  | Ping
+  | Acknowledge
+  | PingRequest of Peer.address
+[@@deriving bin_io]
+
+type config = {
+  protocol_period : int;
+  round_trip_time : int;
+  timeout : int;
+}
+
+type t = {
+  config : config;
+  (* TODO: I am totally not sure about this type *)
+  acknowledges : (int, unit Lwt_condition.t) Poly.t;
+  mutable peers : Peer.t list;
+  (* TODO: I am not sure this is useful, it's just a counter to know what is the current round
+     I would like to send it with the message payload, what do you think? *)
+  mutable sequence_number : int;
+}
+
+(* Helper to increase the round *)
+let next_seq_no t =
+  let sequence_number = t.sequence_number in
+  t.sequence_number <- t.sequence_number + 1;
+  sequence_number
+
+let send_ping_to client (peer_address : Peer.address) =
+  let ping = Util.Encoding.pack bin_writer_message Ping in
+  Client.send_to client ping peer_address
+
+let send_acknowledge_to client (peer_address : Peer.address) =
+  let acknowledge = Util.Encoding.pack bin_writer_message Acknowledge in
+  Client.send_to client acknowledge peer_address
+
+let send_ping_request_to client (peer_address : Peer.address) =
+  let ping_request =
+    Util.Encoding.pack bin_writer_message (PingRequest peer_address) in
+  Client.send_to client ping_request peer_address
+
+(* TODO: Totally not sure about this...
+   I don't know if I can get rid of it and only use `wait_ack_timeout` instead
+    But I feel like Lwt_condition allows ms to manage async thread and could be useful *)
+let rec wait_ack t sequence_number =
+  let cond =
+    find_or_add t.acknowledges sequence_number ~default:Lwt_condition.create
+  in
+  Lwt_condition.wait cond
+
+(* TODO: An async thread should be started here, with a basic sleep, but not sure at all...
+     And we should return Either Ok Timeout *)
+let wait_ack_timeout t sequence_number timeout = failwith "undefined"
+
+(* Basic Knuth shuffle => https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle *)
+let knuth_shuffle known_peers =
+  let initial_array = Array.length (Array.of_list known_peers) in
+  let shuffled_array = Array.copy (Array.of_list known_peers) in
+  for i = initial_array - 1 downto 1 do
+    let k = Random.int (i + 1) in
+    let x = shuffled_array.(k) in
+    shuffled_array.(k) <- shuffled_array.(i);
+    shuffled_array.(i) <- x
+  done;
+  Array.to_list shuffled_array
+
+(* Regarding the SWIM protocol, the list of peers is not ordered.
+   Hence, I basically went for a shuffle after adding the new peer *)
+let add_peer peer_to_add t =
+  let new_peers = peer_to_add :: t.peers in
+  let shuffled_list = knuth_shuffle new_peers in
+  t.peers <- shuffled_list
+
+(* Regarding the SWIM protocol, if peer A cannot get ACK from peer B (timeout):
+   A sets B as `suspicious`
+   A randomly picks one (or several, should it also be randomly determined?) peer(s) from its list
```

Comment:
Number of peers to choose each round can be configurable

Reply:
One way to randomly choose x items from the list without chance of drawing the same item twice is to simply shuffle the list then take the first x items from the shuffled list.

Reply:
Oh yes, I totally forgot that!
Switched to number of peers instead of only one. Thanks.